### PR TITLE
fix: Make supporting classes of AwsCredentials serializable

### LINKED ISSUE #1112: General: Various supporting classes when using AwsCredentials are not serializable
Thanks for stopping by to let us know something could be better!

**PLEASE READ**: If you have a support contract with Google, please create an issue in the [support console](https://cloud.google.com/support/) instead of filing on GitHub. This will ensure a timely response.

Please run down the following list and make sure you've tried the usual "quick fixes":

  - Search the issues already opened: https://github.com/googleapis/google-auth-library-java/issues
  - Check for answers on StackOverflow: http://stackoverflow.com/questions/tagged/google-cloud-platform

If you are still having issues, please include as much information as possible:

#### Environment details

1. Specify the API at the beginning of the title. For example, "BigQuery: ...").
   General, Core, and Other are also allowed as types
2. OS type and version:  Mac OS Venture (13.0.1)
3. Java version: 11
4.  version(s): `main`

#### Steps to reproduce

  1. Using the Flink PubSub connector (patched with the latest version of this library), when passing using external account impersonation to AWS, various classes are unable to be serialized / deserialized by Flink (although this would be true of other runtimes that engage in class serialization, eg Spark, Beam, etc.).

#### Code example

```java
// example
```

#### Stack trace
```
Any relevant stacktrace here.
```

#### External references such as API reference guides

- ?

#### Any additional information below

Following these steps guarantees the quickest resolution possible.

Thanks!

### PR DESCRIPTION
**Pull request by jmahonin**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/google-auth-library-java/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #1112 ☕️

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

### CODE REVIEW

**Comment by davidrabinowitz** [comment]
This can fix https://github.com/GoogleCloudDataproc/spark-bigquery-connector/issues/845 as well. @jmahonin Do you need help with it?

**Comment by jmahonin** [comment]
> This can fix https://github.com/GoogleCloudDataproc/spark-bigquery-connector/issues/845 as well. @jmahonin Do you need help with it?

Just a review at this point!

I can confirm this PR fixes the same serialization issue I observed with Flink. I'm not in a position to test against Spark, but based on past experience, I likewise suspect this will fix it there as well.

**Comment by davidrabinowitz** [comment]
@jmahonin That's fine, I'll test it once released. Can you please rebase the PR so it can be merged?

**Comment by jmahonin** [comment]
Will do, expect a ping on Monday

**Comment by jmahonin** [comment]
@davidrabinowitz Rebased

**Review by TimurSadykov — CHANGES_REQUESTED** [review]
Please add serialize unit tests for affected classes. You can leverage existing helper for it by extending the BaseSerializationTest

**Comment by jmahonin** [comment]
Good call @TimurSadykov , added the tests and discovered `ExternalAccountCredentials.ServiceAccountImpersonationOptions` was not serializable. Added that as well.

There was already a serialization test on ImpersonatedCredentials, though they all use a mock transport factory so there isn't coverage on instantiating the factory in `readObject` (which is true of the existing serialization tests as well).

I didn't add a specific test for the deserialization of `SystemEnvironmentProvider`, it is similarly stubbed out with a `TestEnvironmentProvider` in the tests.

File: oauth2_http/javatests/com/google/auth/oauth2/ExternalAccountCredentialsTest.java [hunk 1]
```diff
@@ -954,6 +956,37 @@ public void getRequestMetadata_withQuotaProjectId() throws IOException {
     assertEquals("quotaProjectId", requestMetadata.get("x-goog-user-project").get(0));
   }
 
+  @Test
+  public void serialize() throws IOException, ClassNotFoundException {
```
Comment (lsirac) [inline oauth2_http/javatests/com/google/auth/oauth2/ExternalAccountCredentialsTest.java:960]: add similar tests in sub classes

Comment (jmahonin) [reply oauth2_http/javatests/com/google/auth/oauth2/ExternalAccountCredentialsTest.java:960]: Will aim to have this updated tomorrow.

**Comment by jmahonin** [comment]
I've updated the unit tests to ensure subclasses of ExternalAccountCredentials have similar `serialize()` tests.

However, the `PluggableAuthCredentials` uses a functional interface `PluggableAuthHandler`, which aren't easily serializable. Given the intent appears to be retrieving tokens using scripts / executables, it's not clear to me if this credential source should be serializable at all. Please advise.

**Review by TimurSadykov — APPROVED** [review]
LGTM

File: oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java [hunk 2]
```diff
@@ -66,8 +67,12 @@
  */
 public abstract class ExternalAccountCredentials extends GoogleCredentials {
 
+  private static final long serialVersionUID = 8049126194174465023L;
```
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java:70]: I think the subclasses should have their own private serialVersionUID.

**Comment by jmahonin** [comment]
This fell off my radar for a bit. I see there's an approval and a request for changes re: subclasses having their own serialVersionUIDs. I'm happy to make the change, but want to confirm first given the PR's been approved.

**Comment by lsirac** [comment]
[jmahonin](https://github.com/jmahonin) please make the change and we will merge the PR, LG otherwise

**Comment by jmahonin** [comment]
> [jmahonin](https://github.com/jmahonin) please make the change and we will merge the PR, LG otherwise

This is done. I didn't include a serialVersionUID for `PluggableAuthCredentials` for the same reason described above here
https://github.com/googleapis/google-auth-library-java/pull/1113#issuecomment-1398489043

**Review by lsirac — APPROVED** [review]

**Pull request merged**